File: go/valid_anagram.go
package main

import "fmt"

func validAnagram(word1, word2 string) bool {
	// if the length is different then it's not a valid anagram
	if len(word1) != len(word2) {
		return false
	}

	wordMap1 := map[string]int{}
	wordMap2 := map[string]int{}

	// creates an object containing the count for each character for word1
	for _, char := range word1 {
		letter := string(char)
		counter := 0

		if counter, found := wordMap1[letter]; found {
			counter++
		}

		wordMap1[letter] = counter
	}

	// creates an object containing the count for each character for word2
	for _, char := range word2 {
		letter := string(char)
		counter := 0

		if counter, found := wordMap2[letter]; found {
			counter++
		}

		wordMap2[letter] = counter
	}

	// loop through the keys of wordMap1
	for k := range wordMap1 {
		// checks if there is a letter key of wordMap1 in wordMap2
		if _, found := wordMap2[k]; !found {
			return false
		}

		// checks if the count of the letter key of wordMap1 in wordMap2 is the same
		if wordMap1[k] != wordMap2[k] {
			return false
		}
	}

	return true
}

func main() {
	fmt.Println(validAnagram("god", "dog"))
	fmt.Println(validAnagram("race", "care"))
	fmt.Println(validAnagram("earth", "heart"))
}

File: js/valid_anagram.js
function validAnagram(arr1, arr2) {
    // if the length is different then it's not a valid anagram
    if (arr1.length !== arr2.length) {
        return false;
    }

    let frequencyCounter1 = {};
    let frequencyCounter2 = {};

    // creates an object containing the count for each character for array1
    for (const val of arr1) {
        frequencyCounter1[val] = (frequencyCounter1[val] || 0) + 1;
    }

    // creates an object containing the count for each character for array2
    for (const val of arr2) {
        frequencyCounter2[val] = (frequencyCounter2[val] || 0) + 1;
    }

    // loop through the keys of array1
    for (const key in frequencyCounter1) {
        // checks if there is a letter key of arr1 in arr2
        if (!(key in frequencyCounter2)) {
            return false;
        }

        // checks if the count of the letter key of arr1 in arr2 is the same
        if (frequencyCounter2[key] !== frequencyCounter1[key]) {
            return false;            
        }
    }

    return true;
}

console.log(validAnagram("god", "dog"));
console.log(validAnagram("race", "care"));
console.log(validAnagram("earth", "heart"));